gpflow/models/mixins.py

```diff
@@ -0,0 +1,54 @@
+import abc
+import collections
+from typing import Callable, Optional, Tuple, TypeVar
+
+
+import tensorflow as tf
+import numpy as np
+
+InputData = tf.Tensor
+OutputData = tf.Tensor
+RegressionData = Tuple[InputData, OutputData]
+Data = TypeVar("Data", RegressionData, InputData)
+
+
+class InternalDataTrainingLossMixin:
+    def training_loss(self):
+        """
+        Specify a training loss for this model
+        """
+        return -self.maximum_a_posteriori_objective()
+
+    def training_loss_closure(self, jit=True) -> Callable[[], tf.Tensor]:
+        if jit:
+            return tf.function(self.training_loss)
+        return self.training_loss
+
+
+class ExternalDataTrainingLossMixin:
+    def training_loss(self, data):
+        """
+        Specify a training loss for this model
+        """
+        return -self.maximum_a_posteriori_objective(data)
+
+    def training_loss_closure(self, data: Data, jit=True) -> Callable[[], tf.Tensor]:
+        training_loss = self.training_loss
+        if jit:
+            training_loss = tf.function(
+                self.training_loss
+            )  # TODO need to add correct input_signature here to allow for differently sized minibatches
+
+        def closure():
+            batch = next(data) if isinstance(data, collections.abc.Iterator) else data
```

Comment:
what happens when `next` raises `StopIteration`? Will the optimizer handle that?

Reply:
good question: i dont know. normally you would use a tf.data.Datatset and set it to loop so you never hit `StopIteration`

Reply:
i think we'd need to be clear on that before we can be confident we support batch processing

Reply:
you need to call `.repeat()` on your dataset (otherwise you can't train properly!) and then it loops forever

Reply:
I think it is going to be fine as it is for now. If you exhaust your training set but insist on keep having more optimisation steps your code crashes :woman_shrugging:

Reply:
how easy is that to do? isn't it just a case of specifying too high a number of iterations? that sounds horrendously fragile